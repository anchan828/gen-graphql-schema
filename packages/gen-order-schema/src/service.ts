import {
  buildASTSchema,
  DirectiveNode,
  DocumentNode,
  EnumTypeDefinitionNode,
  FieldDefinitionNode,
  InputValueDefinitionNode,
  ListTypeNode,
  NamedTypeNode,
  ObjectTypeDefinitionNode,
  parse,
  printSchema,
} from 'graphql';
import { DEFAULT_OPTIONS } from './constants';
import { GenOrderTypesOptions } from './options';
import {
  getDirectives,
  getFieldDefinitions,
  getFieldTypeName,
  getObjectTypeDefinitions,
  isBasicType,
  isEnumType,
} from './utils';

export class GenOrderTypesService {
  genOrderTypes(): string {
    if (!this.hasOrderByDirective()) {
      return printSchema(buildASTSchema(this.documentNode));
    }

    const fields = this.getOrderByFieldDefinitions();

    for (const field of fields) {
      const enumTypeDefinition = this.genOrderByEnumTypeDefinition(field);
      if (!enumTypeDefinition) {
        continue;
      }

      this.appendDefinitionToDocumentNode(enumTypeDefinition);
      this.appendOrderByArgumentToFieldNode(field, enumTypeDefinition);
      this.removeOrderByDirective(field);
      this.removeOrderByIgnoreDirective(field);
    }

    return printSchema(buildASTSchema(this.documentNode));
  }
  private hasOrderByDirective(): boolean {
    const definitions = getObjectTypeDefinitions(this.documentNode);
    for (const definition of definitions) {
      const fields = getFieldDefinitions(definition);

      for (const field of fields) {
        const directives = getDirectives(field);
        if (
          directives.find(
            (d: DirectiveNode) =>
              d.name.value === this.options.orderByDirectiveName,
          )
        ) {
          return true;
        }
      }
    }

    return false;
  }
  private getOrderByFieldDefinitions(): FieldDefinitionNode[] {
    const results: FieldDefinitionNode[] = [];
    const definitions = getObjectTypeDefinitions(this.documentNode);
    for (const definition of definitions) {
      const fields = getFieldDefinitions(definition);

      for (const field of fields) {
        const directives = getDirectives(field);
        if (
          directives.find(
            (d: DirectiveNode) =>
              d.name.value === this.options.orderByDirectiveName,
          )
        ) {
          const { name, isList } = getFieldTypeName(field);
          if (!isList) {
            console.warn(
              `Found order directive, but type of ${
                field.name.value
              } was not ListType. So skip.`,
            );
          } else if (isBasicType(name)) {
            console.warn(
              `Found order directive, but type of ${
                field.name.value
              } was basic types. So skip.`,
            );
          } else {
            results.push(field);
          }
        }
      }
    }
    return results;
  }
  private getObjectTypeDefinition(
    field: FieldDefinitionNode,
  ): ObjectTypeDefinitionNode | undefined {
    const { name } = getFieldTypeName(field);

    return getObjectTypeDefinitions(this.documentNode).find(
      definition => definition.name.value === name,
    );
  }

  private hasOrderByIgnoreDirective(field: FieldDefinitionNode): boolean {
    const directives = getDirectives(field);
    if (
      directives.find(
        (d: DirectiveNode) =>
          d.name.value === this.options.orderByIgnoreDirectiveName,
      )
    ) {
      return true;
    }

    return false;
  }
  private getOrderableFieldNames(
    definition: ObjectTypeDefinitionNode,
  ): string[] {
    const orderableFieldNames: string[] = [];
    for (const field of getFieldDefinitions(definition)) {
      const { name } = getFieldTypeName(field);

      if (
        (isBasicType(name) ||
          isEnumType(this.documentNode, name) ||
          this.options.supportOrderableTypes!.includes(name)) &&
        !this.hasOrderByIgnoreDirective(field)
      ) {
        orderableFieldNames.push(field.name.value);
      }
    }

    return orderableFieldNames;
  }

  private genOrderByEnumTypeDefinition(
    field: FieldDefinitionNode,
  ): EnumTypeDefinitionNode | undefined {
    const type = this.getObjectTypeDefinition(field);
    if (!type) {
      return;
    }
    const { name } = getFieldTypeName(field);
    const orderableFieldNames = this.getOrderableFieldNames(type);
    return {
      kind: 'EnumTypeDefinition',
      name: {
        kind: 'Name',
        value: `${name}${this.options.orderEnumTypeSuffix}`,
      },
      directives: [],
      values: [
        ...Array.prototype.concat.apply(
          [],
          orderableFieldNames.map(orderableFieldName => [
            {
              kind: 'EnumValueDefinition',
              name: {
                kind: 'Name',
                value: `${orderableFieldName}_ASC`,
              },
              directives: [],
            },
            {
              kind: 'EnumValueDefinition',
              name: {
                kind: 'Name',
                value: `${orderableFieldName}_DESC`,
              },
              directives: [],
            },
          ]),
        ),
      ],
    };
  }
  private appendDefinitionToDocumentNode(
    enumTypeDefinition: EnumTypeDefinitionNode,
  ): void {
    Reflect.set(this.documentNode, 'definitions', [
      ...this.documentNode.definitions,
      enumTypeDefinition,
    ]);
  }
  private appendOrderByArgumentToFieldNode(
    field: FieldDefinitionNode,
    enumTypeDefinition: EnumTypeDefinitionNode,
  ): void {
    Reflect.set(field, 'arguments', [
      ...(field.arguments || []),
      this.genOrderByArgument(enumTypeDefinition),
    ]);
  }

  private genOrderByArgument(
    enumType: EnumTypeDefinitionNode,
  ): InputValueDefinitionNode {
    let type: NamedTypeNode | ListTypeNode = {
      kind: 'NamedType',
      name: {
        kind: 'Name',
        value: `${enumType.name.value}`,
      },
    };

    if (this.options.orderByArgumentTypeIsList) {
      type = {
        kind: 'ListType',
        type,
      };
    }

    return {
      kind: 'InputValueDefinition',
      name: {
        kind: 'Name',
        value: this.options.orderByArgumentName!,
      },
      type,
      directives: [],
    };
  }
  private removeOrderByDirective(field: FieldDefinitionNode): void {
    const directives = getDirectives(field);
    Reflect.set(
      field,
      'directives',
      directives.filter(
        directive => directive.name.value !== this.options.orderByDirectiveName,
      ),
    );
  }
  private removeOrderByIgnoreDirective(field: FieldDefinitionNode): void {
    const type = this.getObjectTypeDefinition(field);
    if (!type) {
      return;
    }

    for (const f of getFieldDefinitions(type)) {
      if (!this.hasOrderByIgnoreDirective(f)) {
        continue;
      }
      Reflect.set(
        f,
        'directives',
        getDirectives(f).filter(
          directive =>
            directive.name.value !== this.options.orderByIgnoreDirectiveName,
        ),
      );
    }
  }
  private readonly documentNode: DocumentNode;
  constructor(
    readonly types: string | DocumentNode,
    private readonly options: GenOrderTypesOptions = {},
  ) {
    this.options = Object.assign({}, DEFAULT_OPTIONS, options);
    this.documentNode = typeof types === 'string' ? parse(types) : types;
  }
}
