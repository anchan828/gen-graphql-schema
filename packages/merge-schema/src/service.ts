import {
  getObjectOrUnionTypeDefinitions,
  getObjectTypeDefinitionsFromUnion,
  isUnionType,
} from '@anchan828/gen-graphql-schema-common';
import {
  DocumentNode,
  InterfaceTypeDefinitionNode,
  ObjectTypeDefinitionNode,
  parse,
  StringValueNode,
} from 'graphql';
type TypeName = string;
type Description = string;
export class MergeSchemaService {
  private descriptionMap: Map<TypeName, Description> = new Map<
    TypeName,
    Description
  >();

  private interfaceMap: Map<TypeName, InterfaceTypeDefinitionNode[]> = new Map<
    TypeName,
    InterfaceTypeDefinitionNode[]
  >();

  public cache(type: string | DocumentNode): void {
    if (typeof type === 'string') {
      type = parse(type);
    }
    this.cacheDescriptions(type);
    this.cacheInterfaces(type);
  }
  public apply(type: DocumentNode): void {
    this.applyDescriptions(type);
    this.applyInterfaces(type);
  }
  private cacheInterfaces(type: DocumentNode): void {
    for (const definition of this.getAllObjectDefinitionsByDocumentNode(type)) {
      if (!Array.isArray(definition.interfaces)) {
        continue;
      }
      if (!this.interfaceMap.has(definition.name.value)) {
        this.interfaceMap.set(definition.name.value, []);
      }

      const cachedInterfaces = this.interfaceMap.get(definition.name.value)!;

      for (const interfaceDefinition of definition.interfaces) {
        if (
          !cachedInterfaces.find(
            ci => interfaceDefinition.name.value === ci.name.value,
          )
        ) {
          cachedInterfaces.push(interfaceDefinition);
        }
      }
      this.interfaceMap.set(definition.name.value, cachedInterfaces);
    }
  }
  private cacheDescriptions(type: DocumentNode): void {
    for (const definition of this.getAllObjectDefinitionsByDocumentNode(type)) {
      if (!(definition.description && definition.description.value)) {
        continue;
      }

      let description = definition.description.value;

      if (this.descriptionMap.has(definition.name.value)) {
        description = [
          this.descriptionMap.get(definition.name.value),
          definition.name.value,
        ].join('\n');
      }

      this.descriptionMap.set(definition.name.value, description);
    }
  }

  public applyDescriptions(type: DocumentNode): void {
    for (const definition of this.getAllObjectDefinitionsByDocumentNode(type)) {
      if (this.descriptionMap.has(definition.name.value)) {
        Reflect.set(definition, 'description', {
          kind: 'StringValue',
          value: this.descriptionMap.get(definition.name.value),
        } as StringValueNode);
      }
    }
  }
  public applyInterfaces(type: DocumentNode): void {
    for (const definition of this.getAllObjectDefinitionsByDocumentNode(type)) {
      if (this.interfaceMap.has(definition.name.value)) {
        const interfaces = Object.assign(
          [],
          definition.interfaces || [],
        ) as InterfaceTypeDefinitionNode[];
        const cachedInterfaces = this.interfaceMap.get(definition.name.value);
        if (!Array.isArray(cachedInterfaces)) {
          continue;
        }
        for (const cachedInterface of cachedInterfaces) {
          if (
            !interfaces.find(
              (i: InterfaceTypeDefinitionNode) =>
                i.name.value === cachedInterface.name.value,
            )
          ) {
            interfaces.push(cachedInterface);
          }
        }

        Reflect.set(definition, 'interfaces', interfaces);
      }
    }
  }

  private getAllObjectDefinitionsByDocumentNode(
    documentNode: DocumentNode,
  ): ObjectTypeDefinitionNode[] {
    const definitions: ObjectTypeDefinitionNode[] = [];

    for (const objOrUnion of getObjectOrUnionTypeDefinitions(documentNode)) {
      if (isUnionType(objOrUnion)) {
        definitions.push(
          ...getObjectTypeDefinitionsFromUnion(documentNode, objOrUnion),
        );
      } else {
        definitions.push(objOrUnion);
      }
    }
    const cached: Set<string> = new Set<string>();
    return definitions.filter(def => {
      if (!cached.has(def.name.value)) {
        cached.add(def.name.value);
        return true;
      } else {
        return false;
      }
    });
  }
}
