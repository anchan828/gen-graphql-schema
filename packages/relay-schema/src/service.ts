/* eslint-disable @typescript-eslint/no-non-null-assertion */
import {
  appendDefinitionToDocumentNode,
  getDefinitionByName,
  getDirectives,
  getFieldDefinitionsByDirective,
  getFieldTypeName,
  getObjectTypeDefinitionsFromUnion,
  hasDirectiveInDocumentNode,
  isUnionType,
} from "@anchan828/gen-graphql-schema-common";
import * as deepmerge from "deepmerge";
import {
  DocumentNode,
  FieldDefinitionNode,
  InputValueDefinitionNode,
  InterfaceTypeDefinitionNode,
  ObjectTypeDefinitionNode,
  parse,
  TypeDefinitionNode,
  TypeNode,
} from "graphql";
import { DEFAULT_OPTIONS, DESCRIPTIONS } from "./constants";
import { GenRelayTypesOptions } from "./options";
export class GenRelayTypesService {
  public genRelayTypes(): DocumentNode {
    if (!hasDirectiveInDocumentNode(this.documentNode, this.options.relayDirective!.name!)) {
      return this.documentNode;
    }
    this.genRelayDifinitions();

    const fields = getFieldDefinitionsByDirective(this.documentNode, this.options.relayDirective!.name!).filter(
      (field) => field.type.kind === "ListType",
    );

    for (const field of fields) {
      if (field.type.kind === "NamedType") {
        continue;
      }
      const filedType = getFieldTypeName(field);

      if (
        this.options &&
        this.options.relayCnnectionType &&
        this.options.relayCnnectionType.prefix &&
        filedType.name.startsWith(this.options.relayCnnectionType.prefix)
      ) {
        continue;
      }

      if (
        this.options &&
        this.options.relayCnnectionType &&
        this.options.relayCnnectionType.suffix &&
        filedType.name.endsWith(this.options.relayCnnectionType.suffix)
      ) {
        continue;
      }

      this.addNodeInterface(filedType.name);
      this.genEdgeType(filedType.name);
      this.genConnectionType(filedType.name);
      this.replaceConnectionType(filedType.name, field);
      this.removeRelayDirective(field);
    }
    return this.documentNode;
  }

  private getEdgeTypeName(fieldTypeName: string): string {
    return `${this.options.relayEdgeType!.prefix!}${fieldTypeName}${this.options.relayEdgeType!.suffix}`;
  }

  private getConnectionTypeName(fieldTypeName: string): string {
    return `${this.options.relayCnnectionType!.prefix!}${fieldTypeName}${this.options.relayCnnectionType!.suffix}`;
  }

  private replaceConnectionType(fieldTypeName: string, field: FieldDefinitionNode): void {
    Reflect.set(field, "type", {
      kind: "NonNullType",
      type: {
        kind: "NamedType",
        name: {
          kind: "Name",
          value: this.getConnectionTypeName(fieldTypeName),
        },
      },
    } as TypeNode);
    // before: String, after: String, first: Int, last: Int
    Reflect.set(field, "arguments", [
      ...(field.arguments || []),
      {
        kind: "InputValueDefinition",
        name: {
          kind: "Name",
          value: "before",
        },
        type: {
          kind: "NamedType",
          name: {
            kind: "Name",
            value: "String",
          },
        },
      },
      {
        kind: "InputValueDefinition",
        name: {
          kind: "Name",
          value: "after",
        },
        type: {
          kind: "NamedType",
          name: {
            kind: "Name",
            value: "String",
          },
        },
      },
      {
        kind: "InputValueDefinition",
        name: {
          kind: "Name",
          value: "first",
        },
        type: {
          kind: "NamedType",
          name: {
            kind: "Name",
            value: "Int",
          },
        },
      },
      {
        kind: "InputValueDefinition",
        name: {
          kind: "Name",
          value: "last",
        },
        type: {
          kind: "NamedType",
          name: {
            kind: "Name",
            value: "Int",
          },
        },
      },
    ] as InputValueDefinitionNode[]);
  }

  private genEdgeType(fieldTypeName: string): void {
    if (getDefinitionByName(this.documentNode, this.getEdgeTypeName(fieldTypeName))) {
      return;
    }
    const edgeDefinition = {
      kind: "ObjectTypeDefinition",
      name: {
        kind: "Name",
        value: this.getEdgeTypeName(fieldTypeName),
      },
      description: {
        kind: "StringValue",
        value: DESCRIPTIONS.EDGE_TYPE.TYPE(this.getConnectionTypeName(fieldTypeName)),
      },
    } as ObjectTypeDefinitionNode;
    Reflect.set(edgeDefinition, "fields", [
      {
        kind: "FieldDefinition",

        name: {
          kind: "Name",
          value: "node",
        },
        description: {
          kind: "StringValue",
          value: DESCRIPTIONS.EDGE_TYPE.NODE,
        },
        type: {
          kind: "NonNullType",
          type: {
            kind: "NamedType",
            name: {
              kind: "Name",
              value: fieldTypeName,
            },
          },
        },
      },
      {
        kind: "FieldDefinition",
        name: {
          kind: "Name",
          value: "cursor",
        },
        description: {
          kind: "StringValue",
          value: DESCRIPTIONS.EDGE_INTERFACE.CURSOR,
        },
        type: {
          kind: "NamedType",
          name: {
            kind: "Name",
            value: "String",
          },
        },
      },
    ] as FieldDefinitionNode[]);
    Reflect.set(edgeDefinition, "interfaces", [
      {
        kind: "InterfaceTypeDefinition",
        name: {
          kind: "Name",
          value: this.options.relayEdgeInterface!.name!,
        },
      },
    ]);
    appendDefinitionToDocumentNode(this.documentNode, edgeDefinition);
  }

  private genConnectionType(fieldTypeName: string): void {
    if (getDefinitionByName(this.documentNode, this.getConnectionTypeName(fieldTypeName))) {
      return;
    }
    const connectionDefinition = {
      kind: "ObjectTypeDefinition",

      name: {
        kind: "Name",
        value: this.getConnectionTypeName(fieldTypeName),
      },
      description: {
        kind: "StringValue",
        value: DESCRIPTIONS.CONNECTION_TYPE.TYPE(fieldTypeName),
      },
    } as ObjectTypeDefinitionNode;

    Reflect.set(connectionDefinition, "fields", [
      {
        kind: "FieldDefinition",

        name: {
          kind: "Name",
          value: "totalCount",
        },
        description: {
          kind: "StringValue",
          value: DESCRIPTIONS.CONNECTION_TYPE.TOTAL_COUNT(fieldTypeName),
        },
        type: {
          kind: "NamedType",
          name: {
            kind: "Name",
            value: this.options.relayCnnectionInterface!.totalCountType,
          },
        },
      },
      {
        kind: "FieldDefinition",

        name: {
          kind: "Name",
          value: "edges",
        },
        description: {
          kind: "StringValue",
          value: DESCRIPTIONS.CONNECTION_TYPE.EDGES(this.getEdgeTypeName(fieldTypeName)),
        },
        type: {
          kind: "ListType",
          type: {
            kind: "NamedType",
            name: {
              kind: "Name",
              value: this.getEdgeTypeName(fieldTypeName),
            },
          },
        },
      },
      {
        kind: "FieldDefinition",

        name: {
          kind: "Name",
          value: "pageInfo",
        },
        description: {
          kind: "StringValue",
          value: DESCRIPTIONS.CONNECTION_INTERFACE.PAGE_INFO,
        },
        type: {
          kind: "NonNullType",
          type: {
            kind: "NamedType",
            name: {
              kind: "Name",
              value: this.options.relayPageInfoType!.name!,
            },
          },
        },
      },
    ] as FieldDefinitionNode[]);
    Reflect.set(connectionDefinition, "interfaces", [
      {
        kind: "InterfaceTypeDefinition",
        name: {
          kind: "Name",
          value: this.options.relayCnnectionInterface!.name!,
        },
      },
    ]);
    appendDefinitionToDocumentNode(this.documentNode, connectionDefinition);
  }

  private addNodeInterface(fieldTypeName: string): void {
    const definition = getDefinitionByName(this.documentNode, fieldTypeName) as ObjectTypeDefinitionNode;
    const filedTypes: ObjectTypeDefinitionNode[] = [];
    if (isUnionType(definition)) {
      filedTypes.push(...getObjectTypeDefinitionsFromUnion(this.documentNode, definition));
    } else {
      filedTypes.push(definition);
    }
    for (const fieldType of filedTypes) {
      if (!fieldType || this.hasInterface(fieldType, this.options.relayNodeInterface!.name!)) {
        continue;
      }

      Reflect.set(fieldType, "interfaces", [
        ...(fieldType.interfaces || []),
        {
          kind: "InterfaceTypeDefinition",
          name: {
            kind: "Name",
            value: "Node",
          },
        } as InterfaceTypeDefinitionNode,
      ]);
    }
  }

  private hasInterface(definition: ObjectTypeDefinitionNode, interfaceName: string): boolean {
    return (
      Array.isArray(definition.interfaces) &&
      definition.interfaces.find((i: InterfaceTypeDefinitionNode) => i.name.value === interfaceName)
    );
  }

  private removeRelayDirective(field: FieldDefinitionNode): void {
    const directives = getDirectives(field);
    Reflect.set(
      field,
      "directives",
      directives.filter((directive) => directive.name.value !== this.options.relayDirective!.name!),
    );
  }

  private genRelayDifinitions(): void {
    const difinitions: TypeDefinitionNode[] = [];
    if (!getDefinitionByName(this.documentNode, this.options.relayNodeInterface!.name!)) {
      difinitions.push({
        kind: "InterfaceTypeDefinition",
        name: {
          kind: "Name",
          value: this.options.relayNodeInterface!.name,
        },
        description: {
          kind: "StringValue",
          value: DESCRIPTIONS.NODE_INTERFACE.INTERFACE,
        },
        fields: [
          {
            kind: "FieldDefinition",
            name: {
              kind: "Name",
              value: this.options.relayNodeInterface!.idFieldName,
            },
            description: {
              kind: "StringValue",
              value: DESCRIPTIONS.NODE_INTERFACE.ID,
            },
            type: {
              kind: "NonNullType",
              type: {
                kind: "NamedType",
                name: {
                  kind: "Name",
                  value: "ID",
                },
              },
            },
          },
        ],
      } as InterfaceTypeDefinitionNode);
    }
    if (!getDefinitionByName(this.documentNode, this.options.relayEdgeInterface!.name!)) {
      difinitions.push({
        kind: "InterfaceTypeDefinition",
        name: {
          kind: "Name",
          value: this.options.relayEdgeInterface!.name,
        },
        description: {
          kind: "StringValue",
          value: DESCRIPTIONS.EDGE_INTERFACE.INTERFACE,
        },
        fields: [
          {
            kind: "FieldDefinition",
            name: {
              kind: "Name",
              value: "cursor",
            },
            description: {
              kind: "StringValue",
              value: DESCRIPTIONS.EDGE_INTERFACE.CURSOR,
            },
            type: {
              kind: "NamedType",
              name: {
                kind: "Name",
                value: "String",
              },
            },
          },
        ],
      } as InterfaceTypeDefinitionNode);
    }
    if (!getDefinitionByName(this.documentNode, this.options.relayCnnectionInterface!.name!)) {
      difinitions.push({
        kind: "InterfaceTypeDefinition",
        name: {
          kind: "Name",
          value: this.options.relayCnnectionInterface!.name!,
        },
        description: {
          kind: "StringValue",
          value: DESCRIPTIONS.CONNECTION_INTERFACE.INTERFACE,
        },
        fields: [
          {
            kind: "FieldDefinition",
            name: {
              kind: "Name",
              value: "totalCount",
            },
            description: {
              kind: "StringValue",
              value: DESCRIPTIONS.CONNECTION_INTERFACE.TOTAL_COUNT,
            },
            type: {
              kind: "NamedType",
              name: {
                kind: "Name",
                value: this.options.relayCnnectionInterface!.totalCountType,
              },
            },
          },
          {
            kind: "FieldDefinition",
            name: {
              kind: "Name",
              value: "pageInfo",
            },
            description: {
              kind: "StringValue",
              value: DESCRIPTIONS.CONNECTION_INTERFACE.PAGE_INFO,
            },
            type: {
              kind: "NonNullType",
              type: {
                kind: "NamedType",
                name: {
                  kind: "Name",
                  value: this.options.relayPageInfoType!.name!,
                },
              },
            },
          },
          {
            kind: "FieldDefinition",
            name: {
              kind: "Name",
              value: "edges",
            },
            description: {
              kind: "StringValue",
              value: DESCRIPTIONS.CONNECTION_INTERFACE.EDGES,
            },
            type: {
              kind: "ListType",
              type: {
                kind: "NamedType",
                name: {
                  kind: "Name",
                  value: this.options.relayEdgeInterface!.name!,
                },
              },
            },
          },
        ],
      } as InterfaceTypeDefinitionNode);
    }
    if (!getDefinitionByName(this.documentNode, this.options.relayPageInfoType!.name!)) {
      difinitions.push({
        kind: "ObjectTypeDefinition",
        name: {
          kind: "Name",
          value: this.options.relayPageInfoType!.name!,
        },
        description: {
          kind: "StringValue",
          value: DESCRIPTIONS.PAGE_INFO_TYPE.TYPE,
        },
        fields: [
          {
            kind: "FieldDefinition",
            name: {
              kind: "Name",
              value: "startCursor",
            },
            description: {
              kind: "StringValue",
              value: DESCRIPTIONS.PAGE_INFO_TYPE.START_CURSOR,
            },
            type: {
              kind: "NamedType",
              name: {
                kind: "Name",
                value: "String",
              },
            },
          },
          {
            kind: "FieldDefinition",
            name: {
              kind: "Name",
              value: "endCursor",
            },
            description: {
              kind: "StringValue",
              value: DESCRIPTIONS.PAGE_INFO_TYPE.END_CURSOR,
            },
            type: {
              kind: "NamedType",
              name: {
                kind: "Name",
                value: "String",
              },
            },
          },
          {
            kind: "FieldDefinition",
            name: {
              kind: "Name",
              value: "hasNextPage",
            },
            description: {
              kind: "StringValue",
              value: DESCRIPTIONS.PAGE_INFO_TYPE.HAS_NEXT_PAGE,
            },
            type: {
              kind: "NamedType",
              name: {
                kind: "Name",
                value: "Boolean",
              },
            },
          },
          {
            kind: "FieldDefinition",
            name: {
              kind: "Name",
              value: "hasPreviousPage",
            },
            description: {
              kind: "StringValue",
              value: DESCRIPTIONS.PAGE_INFO_TYPE.HAS_PREVIOUS_PAGE,
            },
            type: {
              kind: "NamedType",
              name: {
                kind: "Name",
                value: "Boolean",
              },
            },
          },
        ],
      } as ObjectTypeDefinitionNode);
    }
    appendDefinitionToDocumentNode(this.documentNode, ...difinitions);
  }

  private readonly documentNode: DocumentNode;

  constructor(readonly types: string | DocumentNode, private readonly options: GenRelayTypesOptions = {}) {
    this.options = deepmerge(DEFAULT_OPTIONS, options);
    this.documentNode = typeof types === "string" ? parse(types) : types;
  }
}
