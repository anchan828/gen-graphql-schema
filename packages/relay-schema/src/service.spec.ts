import { buildASTSchema, parse, printSchema } from 'graphql';
import { GenRelayTypesService } from './service';
describe('GenRelayTypesService', () => {
  it('should return same schema when no where directive', () => {
    const types = `type Test { id: ID }`;
    expect(
      printSchema(
        buildASTSchema(new GenRelayTypesService(types).genRelayTypes()),
      ),
    ).toEqual(printSchema(buildASTSchema(parse(types))));
  });

  it('should return added Where types when has where directive', () => {
    expect(
      printSchema(
        buildASTSchema(
          new GenRelayTypesService(
            [
              `type Test { id: ID! }`,
              `type Query { tests: [Test] @relay, tests2: [Test] @relay}`,
            ].join(`\n`),
          ).genRelayTypes(),
        ),
      ),
    ).toEqual(
      printSchema(
        buildASTSchema(
          parse(
            [
              `type Query { 
                tests( before: String, after: String, first: Int, last: Int): TestConnection
                tests2( before: String, after: String, first: Int, last: Int): TestConnection
               }`,
              `type Test implements Node { id: ID! }`,
              `interface Connection { totalCount: Int, pageInfo: PageInfo!, edges: [Edge] }`,
              `interface Node { id: ID! }`,
              `interface Edge { cursor: String }`,
              `type PageInfo { startCursor: String, endCursor: String, hasNextPage: Boolean, hasPreviousPage: Boolean }`,
              `type TestEdge implements Edge { node: Test, cursor: String}`,
              `type TestConnection implements Connection { totalCount: Int, edges: [TestEdge], pageInfo: PageInfo! }`,
            ].join(`\n`),
          ),
        ),
      ),
    );
  });
});
