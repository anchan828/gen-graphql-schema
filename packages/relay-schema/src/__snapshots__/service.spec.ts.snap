// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenRelayTypesService should return added Where types when has union type 1`] = `
"type Test1 implements Node {
  id: ID!
  name: String!
}

type Test2 implements Node {
  id: ID!
  age: Int!
}

union Test = Test1 | Test2

type Query {
  tests(before: String, after: String, first: Int, last: Int): TestConnection!
  tests2(before: String, after: String, first: Int, last: Int): TestConnection!
}

"""An object with an ID."""
interface Node {
  """ID of the object."""
  id: ID!
}

"""An edge in a connection."""
interface Edge {
  """A cursor for use in pagination."""
  cursor: String
}

"""A connection for relay"""
interface Connection {
  """Identifies the total count of items in the connection."""
  totalCount: Int

  """Information to aid in pagination."""
  pageInfo: PageInfo!

  """A list of edges."""
  edges: [Edge]
}

"""Information about pagination in a connection."""
type PageInfo {
  """When paginating backwards, the cursor to continue."""
  startCursor: String

  """When paginating forwards, the cursor to continue."""
  endCursor: String

  """When paginating forwards, are there more items?"""
  hasNextPage: Boolean

  """When paginating backwards, are there more items?"""
  hasPreviousPage: Boolean
}

"""An edge in a TestConnection."""
type TestEdge implements Edge {
  """The item at the end of the edge."""
  node: Test!

  """A cursor for use in pagination."""
  cursor: String
}

"""The connection type for Test"""
type TestConnection implements Connection {
  """Identifies the total count of Test items in the connection."""
  totalCount: Int

  """A list of TestEdge."""
  edges: [TestEdge]

  """Information to aid in pagination."""
  pageInfo: PageInfo!
}"
`;

exports[`GenRelayTypesService should return added Where types when has where directive 1`] = `
"type Test implements Node {
  id: ID!
}

type Query {
  tests(before: String, after: String, first: Int, last: Int): TestConnection!
  tests2(before: String, after: String, first: Int, last: Int): TestConnection!
}

"""An object with an ID."""
interface Node {
  """ID of the object."""
  id: ID!
}

"""An edge in a connection."""
interface Edge {
  """A cursor for use in pagination."""
  cursor: String
}

"""A connection for relay"""
interface Connection {
  """Identifies the total count of items in the connection."""
  totalCount: Int

  """Information to aid in pagination."""
  pageInfo: PageInfo!

  """A list of edges."""
  edges: [Edge]
}

"""Information about pagination in a connection."""
type PageInfo {
  """When paginating backwards, the cursor to continue."""
  startCursor: String

  """When paginating forwards, the cursor to continue."""
  endCursor: String

  """When paginating forwards, are there more items?"""
  hasNextPage: Boolean

  """When paginating backwards, are there more items?"""
  hasPreviousPage: Boolean
}

"""An edge in a TestConnection."""
type TestEdge implements Edge {
  """The item at the end of the edge."""
  node: Test!

  """A cursor for use in pagination."""
  cursor: String
}

"""The connection type for Test"""
type TestConnection implements Connection {
  """Identifies the total count of Test items in the connection."""
  totalCount: Int

  """A list of TestEdge."""
  edges: [TestEdge]

  """Information to aid in pagination."""
  pageInfo: PageInfo!
}"
`;
